graphbolt/src/cuda/index_select_impl.cu

```diff
@@ -48,6 +53,290 @@ std::pair<torch::Tensor, torch::Tensor> Sort(
+struct ConvertToBytes {
+  const indptr_t* in_degree;
+  __host__ __device__ indptr_t operator()(int64_t i) {
+    return in_degree[i] * sizeof(indices_t);
+  }
+};
+
+template <typename indptr_t, typename indices_t>
+void IndexSelectCSCIndices(
+    const int64_t num_nodes, indices_t* const indices,
+    indptr_t* const sliced_indptr, indptr_t* const sub_indptr,
+    const indptr_t* const in_deg, indices_t* const sub_indices,
+    cudaStream_t stream) {
+  auto allocator = cuda::BuildAllocator();
+  thrust::counting_iterator<int64_t> iota(0);
+
+  auto input_buffer_it = thrust::make_transform_iterator(
+      iota, IteratorFunc<indptr_t, indices_t>{sliced_indptr, indices});
+  auto output_buffer_it = thrust::make_transform_iterator(
+      iota, IteratorFunc<indptr_t, indices_t>{sub_indptr, sub_indices});
+  auto buffer_sizes = thrust::make_transform_iterator(
+      iota, ConvertToBytes<indptr_t, indices_t>{in_deg});
+  constexpr int64_t max_copy_at_once = std::numeric_limits<int32_t>::max();
+  // Performs the copy from indices into sub_indices.
+  for (int64_t i = 0; i < num_nodes; i += max_copy_at_once) {
+    size_t workspace_size = 0;
+    CUDA_CALL(cub::DeviceMemcpy::Batched(
+        nullptr, workspace_size, input_buffer_it + i, output_buffer_it + i,
+        buffer_sizes + i, std::min(num_nodes - i, max_copy_at_once), stream));
+    auto temp = allocator.AllocateStorage<char>(workspace_size);
+    CUDA_CALL(cub::DeviceMemcpy::Batched(
+        temp.get(), workspace_size, input_buffer_it + i, output_buffer_it + i,
+        buffer_sizes + i, std::min(num_nodes - i, max_copy_at_once), stream));
+  }
+}
+
+std::tuple<torch::Tensor, torch::Tensor> IndexSelectCSCImpl(
+    torch::Tensor indptr, torch::Tensor indices, torch::Tensor nodes) {
+  auto allocator = cuda::BuildAllocator();
+  auto stream = c10::cuda::getDefaultCUDAStream();
+
+  const int64_t num_nodes = nodes.size(0);
+
+  // Output indptr for the slice indexed by nodes.
+  torch::Tensor sub_indptr =
+      torch::empty(num_nodes + 1, nodes.options().dtype(indptr.scalar_type()));
+  torch::Tensor sub_indices;
+  AT_DISPATCH_INTEGRAL_TYPES(
+      indptr.scalar_type(), "IndexSelectCSCIndptr", ([&] {
+        using indptr_t = scalar_t;
+        auto [in_deg_ptr, sliced_indptr_ptr] =
+            ComputeDegree(indptr.data_ptr<indptr_t>(), nodes, stream);
+        auto in_deg = in_deg_ptr.get();
+        auto sliced_indptr = sliced_indptr_ptr.get();
+        {  // Compute the output indptr, sub_indptr.
+          size_t workspace_size = 0;
+          CUDA_CALL(cub::DeviceScan::ExclusiveSum(
+              nullptr, workspace_size, in_deg, sub_indptr.data_ptr<indptr_t>(),
+              num_nodes + 1, stream));
+          auto temp = allocator.AllocateStorage<char>(workspace_size);
+          CUDA_CALL(cub::DeviceScan::ExclusiveSum(
+              temp.get(), workspace_size, in_deg,
+              sub_indptr.data_ptr<indptr_t>(), num_nodes + 1, stream));
+        }
```

Comment:
Could you wrap these code as a utility function `cumsum`? Then it is simply `auto sub_indptr = cumsum(in_degree);`.

Reply:
It doesn't take a tensor as input, `in_deg` is a `std::unique_ptr`.

Reply:
In most cases, it may be important to specify the output pointer manually. Otherwise, if we already have a `torch::Tensor`, we can already call `.cumsum()` on it.